0016-base-primitive-unification.md

```diff
@@ -0,0 +1,145 @@
+
+```python
+# prototype implementation for the RFC
+@dataclass(frozen=True)
+class Task:
+    circuit: QuantumCircuit
+```
+
+Different primitive types (such as `Estimator`) are intended to subclass this class, adding auxiliary fields as 
+required.
+
+### DataBin
+
+A data bin is a namespace for storing data.
+The fields in the namespace, in general, depend on the task that was executed (not the `type` of the task).
+For example, a `Sampler` will have fields for each output (as defined by the OpenQASM 3 spec, but in Qiskit, you can currently take "output" to mean the names of the classical registers) of the task's circuit.
+The value of each field will store the corresponding data.
+
+All primitives will store their data in `DataBin`s.
+There will not be subclassing to the effect of `SamplerDataBin < DataBin`; look instead to `TaskResult` for such needs.
+
+```python
+# Make a new DataBin class. make_data_bin() is analagous to dataclasses.make_dataclass().
+# The shape is an optional argument which indicates that all values are to share the same leading shape.
+# To put the following generic example into context, if it helps, imagine that this code lives in the 
+# BaseSampler implementation, and a data bin class is being created to store the results from 
+# particular SamplerTask instance whose circuit has two output registers named alpha and beta, and 
+# that the task itself has shape (5, 4).
+data_bin_cls = make_data_bin({"alpha": NDArray[np.float], "beta": NDArray[np.uint8]}, shape=(5, 4))
+
+# make an instance with particular data
+data_bin = data_bin_cls(
+    alpha=np.empty((5, 4, 1024), dtype=np.float), 
+    beta=np.empty((5, 4, 1024, 127))
+)
+
+# access items as attributes
+alpha_data = data_bin.alpha
+
+# access items with subscripts
+alpha_data = data_bin["alpha"]
+```
+
+### TaskResult
+
+A `TaskResult` is the result of running a single `Task` and does three things:
+
+  * Stores a `DataBin` instance containing the data from execution.
+  * Stores the metadata that is possibly implementation-specific, and always specific to the executed task.
+  * (subclasses) Contains methods to help transform the data into standard formats, including migration helpers.
+
+We generally expect each primitive type to define its own subclass of `TaskResult` to accomodate the third item, though `TaskResult` has no methods that need to be abstract.
+
+We elect to have a special container for the data (`DataBin`) so as not to pollute the `TaskResult` namespace with task-specific names, and to keep data quite distinct from metadata.
+
+```python
+# return a DataBin
+task_result.data
+
+# return a metadata dictionary
+task_result.metadata
+```
+
+
+### PrimitiveResult
+
+`PrimitiveResult` is the type returned by `Job.result()` and is primarily a `Sequence[TaskResult]`, with one entry for each `Task` input into the primitive's run method.
```

Comment:
Super super minor, but Python's `Sequence` requires a few extra methods that we maybe don't want to actually imply - `index` and `count`.

Reply:
That's a good point, I don't want cluttering methods in the classes if we can avoid it. We can chat later about protocols.

Reply:
It's kind of annoying how Sequence is something close but not equal to the thing everyone wants it to be.